File: tests/pytest/test_hhi.py
# tests/pytest/test_hhi.py
import numpy as np
import pytest
from tests.pytest._helpers import _build_df, _extract_period_value, _expected

@pytest.mark.parametrize("period,expected", list(_expected["hhi"].items()))
def test_hhi(period, expected):
    df = _build_df()
    from market_competition_metrics import MarketCompetitionMetrics

    res = MarketCompetitionMetrics.hhi(df, share_col="Market_share", period_col="Period", plot=False)
    value = _extract_period_value(res, period, key_candidates=["HHI","hhi","HHI_t","value","hhi_t"])
    assert np.isclose(value, expected, rtol=1e-3, atol=1e-6), f"HHI mismatch for {period}: got {value}, expected {expected}"

File: tests/testthat/test-hhi.R
library(testthat)

test_that("HHI matches expected per period", {
  df <- build_df()
  res_hhi <- MarketCompetitionMetrics$hhi(
    df,
    share_col = "Market_share",
    period_col = "Period",
    plot = FALSE
  )

  for (p in names(expected_hhi)) {
    val <- extract_period_value(res_hhi, p, key_candidates = c("HHI","hhi"))
    expect_false(is.na(val), info = paste("HHI is NA for period", p))
    expect_equal(as.numeric(val), as.numeric(expected_hhi[p]), tolerance = 1e-6)
  }
})
